clippy_lints/src/option_manual_map.rs

```diff
@@ -0,0 +1,200 @@
+use rustc_lint::{LateContext, LateLintPass, LintContext};
+use rustc_middle::lint::in_external_macro;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::symbol::{sym, Ident};
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for usages of `match` which could be implemented using `Option::map`
+    ///
+    /// **Why is this bad?** Using the `map` method is clearer and more concise.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// match Some(0) {
+    ///     Some(x) => Some(x + 1),
+    ///     None => None,
+    /// };
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// Some(0).map(|x| x + 1);
+    /// ```
+    pub OPTION_MANUAL_MAP,
+    style,
+    "reimplementation of `Option::map`"
+}
+
+declare_lint_pass!(OptionManualMap => [OPTION_MANUAL_MAP]);
+
+impl LateLintPass<'_> for OptionManualMap {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) {
+        if_chain! {
+            if !in_external_macro(cx.sess(), expr.span);
+            if let ExprKind::Match(scrutinee, [arm1, arm2], _) = expr.kind;
+            if is_type_diagnostic_item(cx, cx.typeck_results().expr_ty(expr), sym::option_type);
+            if let Some((some_binding, some_expr, none_expr)) = if is_none_arm(cx, arm1) {
+                get_some_binding(cx, arm2).map(|binding| (binding, arm2.body, arm1.body))
+            } else if is_none_arm(cx, arm2) {
+                get_some_binding(cx, arm1).map(|binding| (binding, arm1.body, arm2.body))
+            } else {
+                None
+            };
+
+            if is_none_expr(cx, none_expr);
+            if let Some(some_expr) = get_some_expr(cx, some_expr);
+
+            then {
+                let mut app = Applicability::MachineApplicable;
+
+                let scrutinee = snippet_with_applicability(cx, scrutinee.span, "_", &mut app);
+                let body = if_chain! {
+                    if let SomeBinding::Name(some_name) = some_binding;
+                    if let ExprKind::Call(func, [arg]) = some_expr.kind;
+                    if let ExprKind::Path(QPath::Resolved(_, arg_path)) = arg.kind;
+                    if let [arg_name] = arg_path.segments;
+                    if arg_name.ident.name == some_name.name;
+                    then {
+                        match func.kind {
+                            ExprKind::Path(QPath::Resolved(_, func_path))
+                                if match_path(func_path, &paths::STD_CONVERT_IDENTITY)
+                                    && !is_allowed(cx, MAP_IDENTITY, expr.hir_id) =>
+                            {
+                                span_lint_and_sugg(
+                                    cx,
+                                    MAP_IDENTITY,
+                                    expr.span,
+                                    "unnecessary map of the identity function",
```

Comment:
I don't think this fits with the `map_identity` lint. `map_identity` is for `.map(|x| x)`. If we're going to lint, `std::convert::identity(_)`, then it should lint *anywhere*, not just in a `match`. In other words, remove this and open a new lint called `invoke_identity`.

Reply:
While I am sympathetic to this change, I'd like to leave it to @Jarcho if he wants to extend the PR to rename the `map_identity` lint. Otherwise we can still do it in a followup PR.

Reply:
It fits the spirit of `map_identity`, and the suggested change in this case would trigger `map_identity`.

i.e.
```rust
match x {
    Some(x) => Some(std::convert::identity(x)),
    None => None,
}
// suggested fix
x.map(std::convert::identity)
```

A lint checking for match expressions that just return the scrutinee would probably be better though.

Reply:
This has been removed.